src/packet/bwe/loss_controller.rs

```diff
@@ -0,0 +1,1296 @@
+        // save our complete observation
+        self.observations[observation.id as usize % self.config.observation_window_size] =
+            observation;
+
+        // renew the partial observation
+        self.partial_observation = PartialObservation::new();
+
+        // calculate upper bound
+        self.cached_instant_upper_bound = Some(self.calculate_instant_upper_bound());
+
+        true
+    }
+
+    fn get_candidates(&self) -> Vec<ChannelParameters> {
+        let mut bandwidths = vec![];
+
+        let current = self.current_estimate.loss_limited_bandwidth;
+
+        for factor in self.config.candidate_factor.iter() {
+            bandwidths.push(factor * current.as_f64());
+        }
+
+        if self.delay_based_estimate.is_valid()
+            && self.config.append_delay_based_estimate_candidate
+            && self.delay_based_estimate > current
+        {
+            bandwidths.push(self.delay_based_estimate.as_f64());
+        }
+
+        let candidate_bandwidth_upper_bound = self.get_candidate_bandwidth_upper_bound().as_f64();
+
+        if self.config.append_acknowledged_rate_candidate && self.acknowledged_bitrate.is_valid() {
+            bandwidths.push(
+                (self.acknowledged_bitrate * self.config.bandwidth_backoff_lower_bound_factor)
+                    .as_f64(),
+            );
+        }
+
+        if self.config.append_delay_based_estimate_candidate
+            && self.delay_based_estimate.is_valid()
+            && self.delay_based_estimate > current
+        {
+            bandwidths.push(
+                (self.delay_based_estimate * self.config.bandwidth_backoff_lower_bound_factor)
+                    .as_f64(),
+            );
+        }
+
+        let mut candidates = Vec::with_capacity(bandwidths.len());
+
+        for bandwidth in bandwidths.iter_mut() {
+            let mut candidate = self.current_estimate;
+            candidate.loss_limited_bandwidth = if self.config.trendline_integration_enabled {
+                bandwidth.min(candidate_bandwidth_upper_bound).into()
+            } else {
+                bandwidth
+                    .min(
+                        self.current_estimate
+                            .loss_limited_bandwidth
+                            .as_f64()
+                            .max(candidate_bandwidth_upper_bound),
+                    )
+                    .into()
+            };
+            candidate.inherent_loss = self.get_feasible_inherent_loss(&candidate);
+            candidates.push(candidate);
+        }
+
+        candidates
+    }
+
+    fn newtons_method_update(&self, channel_parameters: &mut ChannelParameters) {
+        if self.num_observations == 0 {
+            return;
+        }
+
+        for _ in 0..self.config.newton_iterations {
+            let derivatives = self.get_derivatives(channel_parameters);
+            channel_parameters.inherent_loss -=
+                self.config.newton_step_size * (derivatives.0 / derivatives.1);
```

Comment:
it looks like your safe but still pointing out the / 0 risk here

Reply:
If all the observations have no lost or received packets it could happen. I'll add it to the assert below